keps/sig-api-machinery/2608-crd-field-level-versioning/README.md

```diff
@@ -0,0 +1,816 @@
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
+  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
+  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests for meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes adding support for field-level versioning for Custom Resources (CRs).
+When a new field is added to an existing, built-in API type, it is typically gated by
+the addition of a feature gate in [`kube_features.go`](https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go).
+This gate effectively helps control the lifecycle of the feature that is implemented as
+a result of adding this additional field. Currently it is not possible to version CRs in
+a similar manner. This KEP aims to provide a mechanism that allows feature-gate-like
```

Comment:
I'm not sure we *should* version fields in a CR in a similar manner. The built-in approach was sort of a workaround for having to ship go type definitions while controlling what data could be persisted in not-yet-stable fields. CRDs don't have this limitation… you can omitting an unstable field from the stable CRD manifest seems better in every way to me than including the field (and publishing it in openapi) but then forbidding persisting data in it

Reply:
> omitting an unstable field from the stable CRD manifest seems better

for what it's worth, this was the approach sig-networking took with experimental fields in the gateway CRDs... it would be good to get an experience report from them on that approach

Reply:
I agree the field-level deprecation warning bit is the strongest goal of the proposal. That makes a lot more sense to me than adding feature-gated fields to CRDs

Reply:
CRDs don't need anything like the built in resources because you can just update the CRD to make the alpha fields go away.

Reply:
>you can just update the CRD to make the alpha fields go away.

Agree with this, but if this is becoming an official recommendation, it would be really helpful if there was upstream support for this kind of release channel mechanism.

Reply:
Assuming all CRDs are generated from a central set of Go type definitions with `+kubebuilder` annotations, there is currently no way to indicate the stability of a field. If the official upstream answer is to release separate CRDs with different levels of stability, we need a standard way to generate them.

Reply:
This is the official answer, but I don't know what kubebuilder should do--I would make an issue in that repo and discuss there, maybe?